ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducer.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import io.confluent.ksql.rest.entity.CommandId;
+import io.confluent.ksql.rest.server.computation.Command;
+import java.io.Closeable;
+import java.io.IOException;
+
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.eclipse.jetty.util.IO;
+
+
+public interface TransactionalProducer extends Closeable {
```

Comment:
I think this could benefit from a lot more documentation (and merge it with the documentation that's in the implementation class). Namely there's a lot of questions that go un-answered that would be really helpful (at least for me) to review and understand the code.

Overall:
- What is the expected usage pattern for this class?
- Can it be shared across threads?

About `begin`:
- Is this method blocking?
- If a producer on server A calls `begin()` before server B calls `begin()`, what happens to B? Is B still allowed to proceed?

About `send`:
- When is the send resolved?
  - What happens if I call a send without ever calling an abort/commit (perhaps I crash in between)?
  - What happens if I call multiple sends in one transaction?
- What happens if a send fails (i.e. throws an exception)?
- What happens if I call a send before I call begin or after I call commit/abort/close?

About commit/abort:
- What happens if I call commit/abort without calling send or before calling start()?
- When I call abort, what happens to the events I called send for?
- Does calling abort/commit end my transaction? If so what is `close` used for?

Basically I encourage you to think about all of these scenarios (and come up with your own as well) and document them thoroughly. It gets especially complicated when you have multiple servers trying to do these at the same time.

Reply:
I could add it to the code inline, but I'll probably be copying a lot over from the Kafka Transactional Producer documentation. https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

Maybe it would be just easier to link to that documentation? Most of the cases you listed out our documented there.

Reply:
I was also thinking about just switching over to using the KafkaProducer interface for the implementation.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging#KIP-98-ExactlyOnceDeliveryandTransactionalMessaging-ProducerAPIchanges

There are some additional functions in this interface I don't think we currently need though. The only real nuance between the default KafkaProducer interface and the implementation is that we add an additional block in beginTransaction() for the commandRunner to catch up and process all the records. If we still want to have our own interface for this producer, would copying over the Kafka documentation be sufficient?

Reply:
Getting these APIs right is pretty hard - if you think we don't necessarily need to have our own interface, i'd much rather just use the vanilla kafka one. The "waiting" behavior is a little unclear to me (e.g. my questions about `begin` weren't easily answered looking at the KafkaProducer code).

Reply:
We should not do that, as it leaks the fact that Kafka is the underlying queue.

Reply:
If we're adding transactionality, especially tied to the semantics that Kafka exposes, I don't think this is a bad thing.

Reply:
@agavra I added a lot more documentation now to the interface. Lmk if there's anything that I missed. A lot of the edge cases brought up go back to doing the transaction process out of order in some fashion. All the methods will throw a KafkaException if called out of order.

Reply:
Thanks @stevenpyzhang! I'll take a look at the new documentation and add specific comments if I have questions

Reply:
My opinion is that it's worthwhile to keep the rest endpoints decoupled from how we distribute/handle commands. It's already useful for testing, and gives us more flexibility to change things if we want to down the line. I'm not going to die on this hill though. But if we're going that route, I'd like to see two things:

- Just use Kafka's producer interface. There's no value adding and maintaining our own interface here, but plenty of cost (keeping semantics consistent, keeping documentation up to date)
- Let's push the usage of the interface as close to the usage of CommandStore as possible. So create the producer right before beginning each transaction and don't pass it around everywhere (e.g. just create it from `DistributingExecutor.execute`).

Reply:
> Let's push the usage of the interface as close to the usage of CommandStore as possible. So create the producer right before beginning each transaction and don't pass it around everywhere (e.g. just create it from DistributingExecutor.execute).

I guess this also solves the problem of not creating the producer for a set of non-distributed statement. Isn't the overhead going to be a lot for creating a new TransactionalProducer(KafkaProducer and KafkaConsumer) for every single distributed statement?

Reply:
I'm not so worried about overheads for distributed commands - it's probably not perceivable to the human at the end of the CLI, and nobody is submitting these commands programmatically by the thousands

Reply:
If a user were to only ever submit single statements at a time, the overhead is around the same, this would only really be noticeable for large batches of distributed statements.

Reply:
> nobody is submitting these commands programmatically by the thousands

Who knows, given enough users, maybe one of them will come up with some bizarre use case for that. But yea, this makes sense currently. I can refactor to keep everything in the DistributingExecutor.

Reply:
Isn't there a default limit of like 10 running queries at one time anyway? 😂

Reply:
I'd mostly be worried about the added latency. But that would only be for batch statements, which are probably not the norm anyway. I'd favor code quality here unless we discover it's really a problem.

Reply:
Once I make this refactor I can try creating 100 streams in one request

Reply:
100 create streams takes around 10 seconds which is very reasonable.